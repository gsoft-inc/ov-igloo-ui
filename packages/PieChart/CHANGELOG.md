# @igloo-ui/pie-chart

## 0.2.4

### Patch Changes

- cfb5619: Updated hopper token versions and button versions.

## 0.2.3

### Patch Changes

- 7d36815: Updated hopper token versions and igloo icon versions.

## 0.2.2

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency

## 0.2.1

### Patch Changes

- 9739155: Updated storybook and recharts. Applied the new brand look for filter.
- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.

## 0.2.0

### Minor Changes

- 4096378: Added the rebrand. Added disabled prop to Popover to work with StackedBar.

## 0.1.1

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.

## 0.1.0

### Minor Changes

- 89d10c5: Initial release of the pie chart component
