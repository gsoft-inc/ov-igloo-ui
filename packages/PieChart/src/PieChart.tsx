import * as React from 'react';
import cx from 'classnames';
import { PieChart as RPieChart, Cell, Label, Pie } from 'recharts';

import tokens from '@igloo-ui/tokens/dist/base10/tokens.json';

import './pie-chart.scss';

export interface DataProps {
  /** The class name of the data element in the chart */
  className?: string;
  /** The color of the data element */
  color?: string;
  /** The ID of the data element */
  id: string;
  /** The percentage value of the data element */
  percentage: number;
}

export interface PieChartProps extends React.ComponentProps<'div'> {
  /** The class name of the component */
  className?: string;
  /** The data to be displayed in the chart */
  data?: DataProps[];
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** The label displayed in the center of the chart
   * (This is not displayed if the chart size is "regular")
   */
  label?: React.ReactNode;
  /** The rate of the chart in percentage
   * (This is not displayed if the chart size is "regular")
   */
  rate?: number;
  /** The size of the chart */
  size?: 'regular' | 'large';
}

const CustomLabel = (labelObj: any): React.ReactElement => {
  const { viewBox, rate, label } = labelObj;
  const { cx, cy } = viewBox;
  const halfOfSymbolFontSize = 9;
  const halfOfRateHeight = '2.8rem';

  return (
    <text
      x={cx}
      y={cy - 5}
      fill="#3d405c"
      className="recharts-text recharts-label"
      textAnchor="middle"
      dominantBaseline="central"
      alignmentBaseline="middle"
    >
      <tspan
        x={cx + halfOfSymbolFontSize}
        alignmentBaseline="middle"
        className="ids-pie-chart__rate"
      >
        {rate}
      </tspan>
      <tspan
        baselineShift="super"
        alignmentBaseline="middle"
        className="ids-pie-chart__symbol"
      >
        %
      </tspan>
      <tspan x={cx} dy={halfOfRateHeight} className="ids-pie-chart__label">
        {label}
      </tspan>
    </text>
  );
};

const PieChart: React.FunctionComponent<PieChartProps> = ({
  className,
  data = [
    {
      id: 'empty',
      percentage: 100,
      className: 'ids-pie-chart__empty-data',
      color: tokens.grey300,
    },
  ],
  dataTest,
  label,
  rate = 0,
  size = 'large',
}: PieChartProps) => {
  const isEmpty = rate === 0;
  const isLarge = size === 'large';
  const chartSize = size === 'regular' ? 132 : 180;
  const classes = cx('ids-pie-chart', className, {
    'ids-pie-chart--empty': isEmpty,
    'ids-pie-chart--regular': !isLarge,
    'ids-pie-chart--large': isLarge,
  });

  return (
    <div className={classes} data-test={dataTest}>
      <RPieChart width={chartSize} height={chartSize}>
        <Pie
          data={data}
          dataKey="percentage"
          nameKey="id"
          innerRadius={isEmpty && isLarge ? '99%' : '85%'}
          outerRadius="100%"
          isAnimationActive={false}
          startAngle={90}
          endAngle={-270}
        >
          {isLarge && (
            <Label
              width={30}
              position="center"
              content={<CustomLabel rate={rate} label={label} />}
            />
          )}
          {data &&
            data.map((entry) => (
              <Cell
                key={entry.id}
                className={`ids-pie-chart__${entry.id}`}
                fill={entry.color}
                stroke={entry.color}
              />
            ))}
        </Pie>
      </RPieChart>
    </div>
  );
};

export default PieChart;
