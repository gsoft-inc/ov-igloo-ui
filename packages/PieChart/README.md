# PieChart

A pie chart is a circular graph used to display proportions or percentages. It includes a label and percentage in the middle, representing the category or component being depicted and its proportion of the whole.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/pie-chart` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/pie-chart
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/pie-chart
```

## Usage

Then to use the component in your code just import it!

```jsx
import PieChart from '@igloo-ui/pie-chart';
import tokens from '@igloo-ui/tokens/dist/base10/tokens.json';

const mockData = [
  { id: 'positive', percentage: 25, color: tokens.seaweed400 },
  { id: 'negative', percentage: 10, color: tokens.coral400 },
  { id: 'skipped', percentage: 40, color: tokens.creamsicle100 },
  { id: 'unanswered', percentage: 25, color: tokens.grey300 },
];

<PieChart data={mockData} label="Poll Completion" size="large" rate={35} />;
```
