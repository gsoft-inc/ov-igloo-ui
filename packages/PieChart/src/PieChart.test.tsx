/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';

import PieChart from './PieChart';

const setup = (props = {}) => {
  return render(<PieChart dataTest="ids-pie-chart" {...props} />);
};

describe('PieChart', () => {
  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('ids-pie-chart');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should render the provided data elements', () => {
    const data = [
      { id: 'data1', percentage: 30 },
      { id: 'data2', percentage: 20 },
      { id: 'data3', percentage: 50 },
    ];
    setup({ data });

    const dataElements = screen.getAllByRole('img');
    expect(dataElements.length).toBe(data.length);
  });

  test('It should render the provided label in the center', () => {
    const label = 'Sample Label';
    setup({ label });

    const labelElement = screen.getByText(label);
    expect(labelElement).toBeInTheDocument();
  });

  test('It should render the provided rate in the center', () => {
    const rate = 75;
    setup({ rate });

    const rateElement = screen.getByText(rate);
    expect(rateElement).toBeInTheDocument();
  });

  test('It should render a regular-sized chart when size is "regular"', () => {
    setup({ size: 'regular' });

    const chartContainer = screen.getByTestId('ids-pie-chart');
    expect(chartContainer).toHaveClass('ids-pie-chart--regular');
  });

  test('It should render a large-sized chart when size is "large"', () => {
    setup({ size: 'large' });

    const chartContainer = screen.getByTestId('ids-pie-chart');
    expect(chartContainer).toHaveClass('ids-pie-chart--large');
  });

  test('It should render the default "empty" data if no data is provided', () => {
    const { container } = setup({ data: undefined });

    const emptyDataElement = container.querySelector('.ids-pie-chart--empty');
    expect(emptyDataElement).toBeInTheDocument();
  });
});
