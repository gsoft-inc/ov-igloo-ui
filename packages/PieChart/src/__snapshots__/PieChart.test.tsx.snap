// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-pie-chart ids-pie-chart--empty ids-pie-chart--large"
    data-test="ids-pie-chart"
  >
    <div
      class="recharts-wrapper"
      role="region"
      style="position: relative; cursor: default; width: 180px; height: 180px;"
    >
      <svg
        class="recharts-surface"
        cx="50%"
        cy="50%"
        height="180"
        viewBox="0 0 180 180"
        width="180"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts3-clip"
          >
            <rect
              height="170"
              width="170"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
          tabindex="0"
        >
          <g
            class="recharts-layer recharts-pie-sector"
            tabindex="-1"
          >
            <path
              class="recharts-sector ids-pie-chart__empty"
              color="#D5D9DE"
              cx="90"
              cy="90"
              d="M 90,5
    A 85,85,0,
    1,1,
    89.99851647013588,5.000000012946245
  L 89.99853130543453,5.850000012816778
            A 84.15,84.15,0,
            1,0,
            90,5.849999999999994 Z"
              fill="#D5D9DE"
              id="empty"
              name="empty"
              role="img"
              stroke="#D5D9DE"
              tabindex="-1"
            />
          </g>
          <text
            alignment-baseline="middle"
            class="recharts-text recharts-label"
            dominant-baseline="central"
            fill="#3d405c"
            text-anchor="middle"
            x="90"
            y="85"
          >
            <tspan
              alignment-baseline="middle"
              class="ids-pie-chart__rate"
              x="99"
            >
              0
            </tspan>
            <tspan
              alignment-baseline="middle"
              baseline-shift="super"
              class="ids-pie-chart__symbol"
            >
              %
            </tspan>
            <tspan
              class="ids-pie-chart__label"
              dy="2.8rem"
              x="90"
            />
          </text>
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
